index.js

```diff
@@ -124,6 +151,298 @@
+      return {
+        // before self
+        open: open,
+        // after self
+        close: close,
+        // before child
+        prefix: prefix,
+        // between siblings
+        glue: glue,
+        // after close
+        separator: separator
+      };
+    }
+    
+    var grammars = {
+      SIG:         Grammar('' , '' , '' , ''     , ''    ),
+      
+      NAME:        Grammar('' , '' , '' , ''     , ' :: '),
+      CONSTRAINTS: Grammar('' , '' , '' , ', '   , ' => '),
+      ARGS:        Grammar('' , '' , '' , ' -> ' , ' -> '),
+      RET:         Grammar('' , '' , '' , ''     , ''    ),
+      
+      VARIABLE:    Grammar('' , '' , '' , ' '    , ''    ),
+      UNARY:       Grammar('(', ')', ' ', ' '    , ''    ),
+      BINARY:      Grammar('(', ')', ' ', ' '    , ''    ),
+      
+      DELIMITER:   Grammar('' , '' , '' , ''     , ''    ),
+      
+      // Number, String, ValidDate, etc
+      _DEFAULT:    Grammar('' , '' , '' , ''     , ''    ),
+    };
+    
+    // depends on "global" grammars
+    // Nullable b/c parent will be null on traverses that start at root
+    //  rule :: Nullable Object -> String
+    var rule = function(token, key) {
+      return (
+        token == null || !has(token.type, grammars) ? grammars._DEFAULT[key] :
+        grammars[token.type][key]
+      );
+    }
+    
+    //  Token :: (String, String, [Token]) -> Token
+    var Token = function(type, text, children) {
+      return {
+        type: type,
+        text: text,
+        start: null,
+        end: null,
+        tokens: children
+      };
+    };
+    
+    //  Name :: String -> Token
+    var Name = function(data) {
+      return Token('NAME', '', [
+        Token('FN_NAME', data, [])
+      ]);
+    };
+    
+    //  Constraints :: [Constraint] -> Token
+    var Constraints = function(data) {
+      return Token('CONSTRAINTS', '', []);
+    };
+    
+    //  Args :: [ExpType] -> Token
+    var Args = function(data) {
+      return Token('ARGS', '', map(data, Arg));
+    };
+    
+    //  Ret :: ExpType -> Token
+    var Ret = function(data) {
+      return Token('RET', '', [
+        Arg(data)
+      ]);
+    };
+    
+    //  Arg :: ExpType -> Token
+    var Arg = function(arg) {
+      return Token(
+        arg.type,
+        arg.name ? stripNamespace(arg.name) : arg.toString(),
+        map(getArgChildren(arg), Arg));
+    };
+    
+    //  getArgChildren :: ExpType -> [ExpType]
+    var getArgChildren = function(arg) {
+      switch (arg.type) {
+        case 'UNARY':  return [arg.$1];
+        case 'BINARY': return [arg.$1, arg.$2];
+        default:       return [];
+      }
+    };
+    
+    //  insert :: [Token] -> Token -> Token -> Integer -> Void
+    var insert = function(children, target, newChild, offset) {
+      for (var idx = 0; idx < children.length; idx += 1) {
+        if (children[idx] === target) {
+          children.splice(idx + offset, 0, newChild);
```

Comment:
I'm sure that it is possible ;) Don't know how easy/difficult though. I made a comment elsewhere related to this.. we create a cloned array as we traverse, which simplifies things I think. I'm not really sure what we would gain, we would just end up changing
```js

      traverse(addPrefix, null, sig);
      traverse(addWrapper, null, sig);
      traverse(addSeparator, null, sig);
      pairTraverse(addGlue, null, sig);
```
to
```js
      sig = traverse(addPrefix, null, sig);
      sig = traverse(addWrapper, null, sig);
      sig = traverse(addSeparator, null, sig);
      sig = pairTraverse(addGlue, null, sig);
```
or a nicer looking version that uses a pipe of some sort to achieve the same thing. If we get everything to a place we like and this becomes a sticking point, I can revisit it. I know babel plugins mutate their AST as they traverse. And sanctuary-def does some mutation already as well https://github.com/plaid/sanctuary-def/blob/master/index.js#L762